### PR TITLE
Refactor/preparep api/function like opcodes

### PR DESCRIPTION
**Pull request by DemuirGos**
Fixes | Closes | Resolves #

> Anything marked as optional that you didn't need to fill in, please remove it from the PR description. Choose one of the keywords above to refer to the issue this PR solves, followed by the issue number (e.g Fixes # 666). If there is no issue, remove the line. Remove this note after reading.

## Changes:
- Added Extension Methods to simplify Preparing bytecode tests using Prepare API

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm.Test/FluentByteCodeBuildeTests.cs [hunk 1]
```diff
@@ -0,0 +1,456 @@
+//  Copyright (c) 2021 Demerzel Solutions Limited
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.Evm.Test/FluentByteCodeBuildeTests.cs:1]: The filename is not right.

File: src/Nethermind/Nethermind.Evm/ByteCodeBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+using Org.BouncyCastle.Asn1.Mozilla;
+
+namespace Nethermind.Evm
+{
+    /// <summary>
+    /// A utility class that extends Prepare class to abstract stack-like behaviour
+    /// </summary>
+    ///
+
+    //Note(Ayman) : redesign arg order to add flexibility and ability to use args already in stack, and add tests
```
Comment (smartprogrammer93) [inline src/Nethermind/Nethermind.Evm/ByteCodeBuilderExtensions.cs:34]: I believe that is already done. so delete the comment

**Review by smartprogrammer93 — CHANGES_REQUESTED** [review]
overall this is great. just one fileaneme fix and its mergable

File: src/Nethermind/Nethermind.Evm/ByteCodeBuilderExtensions.cs (hunk 2 above)
Comment (DemuirGos) [reply src/Nethermind/Nethermind.Evm/ByteCodeBuilderExtensions.cs]: Yes it was resolved.

**Review by smartprogrammer93 — APPROVED** [review]
LGTM

**Review by LukaszRozmej — COMMENTED** [review]
Why use nullables and later ignore them? So if someone uses nulls to those methods than will get bad results?

**Comment by DemuirGos** [comment]
> Why use nullables and later ignore them? So if someone uses nulls to those methods than will get bad results?

null arguments don't get emitted, the functions use these functions internal ([link](https://github.com/NethermindEth/nethermind/blob/424f9615bd8ad04e5dc95d273d403561d2733cb4/src/Nethermind/Nethermind.Evm/ByteCodeBuilderExtensions.cs#L35)):
```
internal static Prepare PushSingle(this Prepare @this, UInt256? value);
internal static Prepare PushSingle(this Prepare @this, byte[] value);
internal static Prepare PushSingle(this Prepare @this, Address? value);
``` 
instead of `PushData` directly, because in some cases we would want to re-use arguments from the stack, like this :
```
PUSH1 1
PUSH1 2
ADD
PUSH1 3
SUB
```
this would be in the function-opcode : 
```CSHARP
EvmCode
.ADD(1, 2)
.SUB(3, null) // <= here we passed second argument as null in order to reuse ADD instruction result
.Done
```
that's why most arguments are defaulted to `null`.

**Pull request merged**